File: fp-cpp/lib/Matrix/Multiply.cpp
#include "Matrix/Multiply.h"
#include "Util/Math.h"

#include <cassert>
#include <thread>

using namespace matrix;

Dense matrix::multiply(const Dense &A, const Dense &B) {
  auto [ARows, ACols] = A.dim();
  auto [BRows, BCols] = B.dim();
  assert(ACols == BRows && "mismatch of dimensions");

  Dense CMat(ARows, BCols);
  for (IndexT R = 0; R < ARows; ++R) {
    for (IndexT C = 0; C < BCols; ++C) {
      double Sum = 0.0;
      for (IndexT K = 0; K < ACols; ++K)
        Sum += A.at(R, K) * B.at(K, C);
      CMat.at(R, C) = Sum;
    }
  }
  return CMat;
}

Dense matrix::multiplyBlocked(const Dense &A, const Dense &B) {
  auto [ARows, ACols] = A.dim();
  auto [BRows, BCols] = B.dim();
  assert(ACols == BRows && "mismatch of dimensions");

  Dense CMat(ARows, BCols);
  constexpr IndexT BlockSize = 8;
  for (IndexT RO = 0; RO < ARows; RO += BlockSize) {
    for (IndexT CO = 0; CO < BCols; CO += BlockSize) {
      for (IndexT KO = 0; KO < ACols; KO += BlockSize) {
        for (IndexT R = RO; R < std::min(RO + BlockSize, ARows); ++R) {
          for (IndexT C = CO; C < std::min(CO + BlockSize, BCols); ++C) {
            double Sum = KO == 0 ? 0.0 : CMat.at(R, C);
            for (IndexT K = KO; K < std::min(KO + BlockSize, ACols); ++K)
              Sum += A.at(R, K) * B.at(K, C);
            CMat.at(R, C) = Sum;
          }
        }
      }
    }
  }
  return CMat;
}

Dense matrix::multiplyParallel(const Dense &A, const Dense &B) {
  IndexT ARows, ACols, BRows, BCols;
  std::tie(ARows, ACols) = A.dim();
  std::tie(BRows, BCols) = B.dim();
  assert(ACols == BRows && "mismatch of dimensions");

  constexpr IndexT BlockSize = 8;
  IndexT MaxProc = std::thread::hardware_concurrency();
  IndexT WorkWidth =
      util::divRU(util::divRU(ARows, MaxProc), BlockSize) * BlockSize;

  Dense CMat(ARows, BCols);
  std::vector<std::thread> Thrds(MaxProc);

  for (IndexT TID = 0; TID < MaxProc && TID * WorkWidth < ARows; ++TID)
    Thrds[TID] = std::thread(
        [WorkWidth, ARows, ACols, BCols, &A, &B, &CMat](IndexT Start) {
          for (IndexT R = Start; R < std::min(Start + WorkWidth, ARows); ++R) {
            for (IndexT C = 0; C < BCols; ++C) {
              double Sum = 0.0;
              for (IndexT K = 0; K < ACols; ++K)
                Sum += A.at(R, K) * B.at(K, C);
              CMat.at(R, C) = Sum;
            }
          }
        },
        TID * WorkWidth);

  for (auto &Thrd : Thrds)
    Thrd.join();

  return CMat;
}

Dense matrix::multiplyFast(const Dense &A, const Dense &B) {
  IndexT ARows, ACols, BRows, BCols;
  std::tie(ARows, ACols) = A.dim();
  std::tie(BRows, BCols) = B.dim();
  assert(ACols == BRows && "mismatch of dimensions");

  constexpr IndexT BlockSize = 8;
  IndexT MaxProc = std::thread::hardware_concurrency();
  IndexT WorkWidth =
      util::divRU(util::divRU(ARows, MaxProc), BlockSize) * BlockSize;

  Dense CMat(ARows, BCols);
  std::vector<std::thread> Thrds(MaxProc);

  for (IndexT TID = 0; TID < MaxProc && TID * WorkWidth < ARows; ++TID)
    Thrds[TID] = std::thread(
        [WorkWidth, ARows, ACols, BCols, &A, &B, &CMat](IndexT Start) {
          for (IndexT RO = Start; RO < std::min(Start + WorkWidth, ARows);
               RO += BlockSize) {
            for (IndexT CO = 0; CO < BCols; CO += BlockSize) {
              for (IndexT KO = 0; KO < ACols; KO += BlockSize) {
                for (IndexT R = RO; R < std::min(RO + BlockSize, ARows); ++R) {
                  for (IndexT C = CO; C < std::min(CO + BlockSize, BCols);) {
                    if (C + 4 <= std::min(CO + BlockSize, BCols)) {
                      std::array<double, 4> Sum;
                      for (IndexT I = 0; double &S : Sum)
                        S = KO == 0 ? 0.0 : CMat.at(R, C + (I++));
                      for (IndexT K = KO; K < std::min(KO + BlockSize, ACols);
                           ++K)
                        for (IndexT I = 0; double &S : Sum)
                          S += A.at(R, K) * B.at(K, C + (I++));
                      for (IndexT I = 0; double &S : Sum)
                        CMat.at(R, C + (I++)) = S;
                      C += 4;
                    } else {
                      double Sum = KO == 0 ? 0.0 : CMat.at(R, C);
                      for (IndexT K = KO; K < std::min(KO + BlockSize, ACols);
                           ++K)
                        Sum += A.at(R, K) * B.at(K, C);
                      CMat.at(R, C) = Sum;
                      ++C;
                    }
                  }
                }
              }
            }
          }
        },
        TID * WorkWidth);

  for (IndexT TID = 0; TID < MaxProc && TID * WorkWidth < ARows; ++TID)
    Thrds[TID].join();

  return CMat;
}

File: fp-rust/src/matrix/multiply.rs
use super::*;
use crate::util;
use crossbeam::thread;
use std::cmp::min;

pub fn multiply(a: &Matrix, b: &Matrix) -> Matrix {
    let (a_row, a_col) = a.dim();
    let (b_row, b_col) = b.dim();
    if a_col != b_row {
        panic!(
            "dimenions do not match: ({}x{}) ({}x{})",
            a_row, a_col, b_row, b_col
        );
    }
    let mut c_mat = Matrix::new(a_row, b_col);
    for r in 0..a_row {
        for c in 0..b_col {
            let mut sum = 0.0;
            for k in 0..a_col {
                sum += a.at(r, k) * b.at(k, c);
            }
            c_mat.set(r, c, sum);
        }
    }
    c_mat
}

pub fn multiply_blocked(a: &Matrix, b: &Matrix) -> Matrix {
    let (a_row, a_col) = a.dim();
    let (b_row, b_col) = b.dim();
    if a_col != b_row {
        panic!(
            "dimenions do not match: ({}x{}) ({}x{})",
            a_row, a_col, b_row, b_col
        );
    }
    let mut c_mat = Matrix::new(a_row, b_col);
    const BLOCK_SIZE: usize = 8;
    for ro in (0..a_row).step_by(BLOCK_SIZE) {
        for co in (0..b_col).step_by(BLOCK_SIZE) {
            for ko in (0..a_col).step_by(BLOCK_SIZE) {
                for r in ro..min(ro + BLOCK_SIZE, a_row) {
                    for c in co..min(co + BLOCK_SIZE, b_col) {
                        let mut sum = if ko == 0 { 0.0 } else { c_mat.at(r, c) };
                        for k in ko..min(ko + BLOCK_SIZE, a_col) {
                            sum += a.at(r, k) * b.at(k, c);
                        }
                        c_mat.set(r, c, sum);
                    }
                }
            }
        }
    }
    c_mat
}

pub fn multiply_parallel(a: &Matrix, b: &Matrix) -> Matrix {
    let (a_row, a_col) = a.dim();
    let (b_row, b_col) = b.dim();
    if a_col != b_row {
        panic!(
            "dimenions do not match: ({}x{}) ({}x{})",
            a_row, a_col, b_row, b_col
        );
    }
    let mut c_mat = Matrix::new(a_row, b_col);
    let max_proc = std::thread::available_parallelism().unwrap().get();
    let work_width = util::div_ru(a_row, max_proc);

    thread::scope(|s| {
        let mut rest = c_mat.data.as_mut_slice();
        for start in (0..a_row).step_by(work_width) {
            let (slice, remain) = if start + work_width <= a_row {
                rest.split_at_mut(work_width * b_col)
            } else {
                (rest, [].as_mut_slice())
            };
            rest = remain;
            s.spawn(move |_| {
                for r in start..min(start + work_width, a_row) {
                    for c in 0..b_col {
                        let mut sum = 0.0;
                        for k in 0..a_col {
                            sum += a.at(r, k) * b.at(k, c);
                        }
                        slice[(r - start) * b_col + c] = sum;
                    }
                }
            });
        }
    })
    .unwrap();
    c_mat
}

pub fn multiply_fast(a: &Matrix, b: &Matrix) -> Matrix {
    let (a_row, a_col) = a.dim();
    let (b_row, b_col) = b.dim();
    if a_col != b_row {
        panic!(
            "dimenions do not match: ({}x{}) ({}x{})",
            a_row, a_col, b_row, b_col
        );
    }
    let mut c_mat = Matrix::new(a_row, b_col);

    const BLOCK_SIZE: usize = 8;
    let max_proc = std::thread::available_parallelism().unwrap().get();
    let work_width = util::div_ru(util::div_ru(a_row, max_proc), BLOCK_SIZE) * BLOCK_SIZE;

    thread::scope(|s| {
        let mut rest = c_mat.data.as_mut_slice();
        for start in (0..a_row).step_by(work_width) {
            let (slice, remain) = if start + work_width <= a_row {
                rest.split_at_mut(work_width * b_col)
            } else {
                (rest, [].as_mut_slice())
            };
            rest = remain;
            s.spawn(move |_| {
                let idx = |r, c| (r - start) * b_col + c;
                for ro in (start..min(start + work_width, a_row)).step_by(BLOCK_SIZE) {
                    for co in (0..b_col).step_by(BLOCK_SIZE) {
                        for ko in (0..a_col).step_by(BLOCK_SIZE) {
                            for r in ro..min(ro + BLOCK_SIZE, a_row) {
                                let mut c = co;
                                while c < min(co + BLOCK_SIZE, b_col) {
                                    if c + 4 <= min(co + BLOCK_SIZE, b_col) {
                                        let mut sum = [0.0; 4];
                                        if ko != 0 {
                                            for i in 0..4 {
                                                sum[i] = slice[idx(r, c + i)];
                                            }
                                        }
                                        for k in ko..min(ko + BLOCK_SIZE, a_col) {
                                            for i in 0..4 {
                                                sum[i] += a.at(r, k) * b.at(k, c + i);
                                            }
                                        }
                                        for i in 0..4 {
                                            slice[idx(r, c + i)] = sum[i];
                                        }
                                        c += 4;
                                    } else {
                                        let mut sum = if ko == 0 { 0.0 } else { slice[idx(r, c)] };
                                        for k in ko..min(ko + BLOCK_SIZE, a_col) {
                                            sum += a.at(r, k) * b.at(k, c);
                                        }
                                        slice[idx(r, c)] = sum;
                                        c += 1;
                                    }
                                }
                            }
                        }
                    }
                }
            });
        }
    })
    .unwrap();
    c_mat
}

#[cfg(test)]
mod test {
    use super::*;
    use rand::{self, distributions::Standard, Rng};

    #[test]
    fn multiply_test() {
        for i in 1..101 {
            let a = Matrix::identity(i);
            let b = Matrix::from_vec(i, i, (0..).take(i * i).map(f64::from).collect());
            assert_eq!(b, multiply(&a, &b));
        }
    }

    type MultFunc = fn(&Matrix, &Matrix) -> Matrix;
    fn compare_multiplies(repeat: usize, mult_base: MultFunc, mult_test: MultFunc) {
        for _ in 0..repeat {
            let mut r = rand::thread_rng();
            let (m, k, n) = (
                r.gen_range(10..500),
                r.gen_range(10..500),
                r.gen_range(10..500),
            );
            let nums = r
                .sample_iter(Standard)
                .take(m * k + k * n)
                .collect::<Vec<f64>>();
            let (va, vb) = nums.split_at(m * k);
            let a = Matrix::from_vec(m, k, va.to_vec());
            let b = Matrix::from_vec(k, n, vb.to_vec());
            assert_eq!(mult_base(&a, &b), mult_test(&a, &b));
        }
    }

    #[test]
    fn multiply_blocked_test() {
        compare_multiplies(10, multiply, multiply_blocked);
    }

    #[test]
    fn multiply_parallel_test() {
        compare_multiplies(50, multiply_blocked, multiply_parallel);
    }

    #[test]
    fn multiply_fast_test() {
        compare_multiplies(50, multiply_blocked, multiply_fast);
    }
}
